File: 백준/Silver/15654. N과 M （5）/N과 M （5）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class Main {
	public static void main(String[] args) throws IOException {
		input();
		solution();
	}
	private static int N, M;
	private static int[] arr;
	private static List<String> list;
	private static void solution() {
		int[] res = new int[M];
		boolean[] visited = new boolean[N];
		list = new ArrayList<>();
		perm(res, visited, 0);
		System.out.println(String.join("\n", list));
	}

	private static void perm(int[] res, boolean[] visited, int depth) {
		if (depth == M) {
			list.add(
				Arrays.stream(res).mapToObj(String::valueOf).collect(Collectors.joining(" "))
			);

			return;
		}

		for (int i = 0; i < N; i++) {
			if (visited[i]) continue;

			visited[i] = true;
			res[depth] = arr[i];
			perm(res, visited, depth + 1);
			visited[i] = false;
		}
	}

	private static void input() throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int[] nm = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).sorted().toArray();
		N = nm[0];
		M = nm[1];
	}
}

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.kt
fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readln().split(" ").map { it.toInt() }
    val arr = readln().split(" ").map { it.toInt() }.sorted().toIntArray()
    val visited = BooleanArray(n)
    val result = IntArray(m) // 결과를 저장할 배열 추가
    close()

    perm(arr, visited, result, m, 0)
}

private fun perm(arr: IntArray, visited: BooleanArray, result: IntArray, m: Int, depth: Int) {
    if (depth == m) {
        print(result)
        return
    }

    for (i in arr.indices) {
        if (visited[i]) continue

        visited[i] = true
        result[depth] = arr[i] // 원본 배열 대신 결과 배열에 값 저장
        perm(arr, visited, result, m, depth + 1)
        visited[i] = false
    }
}

private fun print(arr: IntArray) =
    arr.joinToString(" ")
        .let(::println)